File: tools/session-state/src/session_state/capture_scratch.py
"""capture_scratch.py — Stop-hook entry: overwrite this session's scratch with freshest git facts."""
from __future__ import annotations

import json
import os
import sys
from datetime import datetime, timezone
from pathlib import Path

for _s in (sys.stdout, sys.stderr):
    try:
        _s.reconfigure(encoding="utf-8")  # type: ignore[attr-defined]
    except Exception:
        pass

THROTTLE_SECONDS = 30


def main() -> int:
    try:
        if os.environ.get("CC_SESSION_STATE_DISABLE"):
            return 0
        raw = sys.stdin.read() or "{}"
        data = json.loads(raw) if raw.strip() else {}
        session_id = data.get("session_id") or "unknown"

        from session_state import keying, gitfacts, store
        cwd = Path(os.getcwd())
        root = keying.repo_root(cwd)
        repo = str(root) if root else str(cwd)
        dir = keying.state_dir(cwd)
        if not keying.check_meta(dir, repo):
            return 0  # collision: refuse loudly (marker written), don't guess

        git = gitfacts.collect_git_facts(cwd)
        # throttle: skip if recent AND git facts unchanged
        prev = store.scratch_path(dir, session_id)
        if prev.exists():
            try:
                old = json.loads(prev.read_text(encoding="utf-8"))
                age = (datetime.now(timezone.utc)
                       - datetime.fromisoformat(old["ts"].replace("Z", "+00:00"))).total_seconds()
                if age < THROTTLE_SECONDS and old.get("git") == git:
                    return 0
            except Exception:
                pass
        rec = store.make_record("auto", "Stop", session_id, repo, git)
        store.write_scratch(dir, rec)
    except Exception:
        pass
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: plugins/session-state/src/capture_scratch.mjs
import fs from "node:fs";
import * as keying from "./keying.mjs";
import * as gitfacts from "./gitfacts.mjs";
import * as store from "./store.mjs";

const THROTTLE_MS = 30000;
function readStdin() { try { return fs.readFileSync(0, "utf8"); } catch { return ""; } }

try {
  if (!process.env.CC_SESSION_STATE_DISABLE) {
    const raw = readStdin();
    const data = raw.trim() ? JSON.parse(raw) : {};
    const sessionId = data.session_id || "unknown";
    const cwd = process.cwd();
    const root = keying.repoRoot(cwd);
    const repo = root || cwd;
    const dir = keying.stateDir(cwd);
    if (keying.checkMeta(dir, repo)) {
      const git = gitfacts.collectGitFacts(cwd);
      const prev = store.scratchPath(dir, sessionId);
      let skip = false;
      if (fs.existsSync(prev)) {
        try {
          const old = JSON.parse(fs.readFileSync(prev, "utf8"));
          const age = Date.now() - new Date(old.ts).getTime();
          if (age < THROTTLE_MS && JSON.stringify(old.git) === JSON.stringify(git)) skip = true;
        } catch {}
      }
      if (!skip) store.writeScratch(dir, store.makeRecord("auto", "Stop", sessionId, repo, git));
    }
  }
} catch {}
process.exit(0);
